### PR TITLE
Batch custom scopes activation

### PR DESCRIPTION
**Pull request by luca-bassoricci**
- Fixes https://github.com/quarkiverse/quarkus-jberet/issues/122

Added job/step/partition scope management

I've added Quarkus specific implementation for batch scopes in runtime module and registered them in `JBeretProcessor`.
I marked PR as draft mainly because:

1. I'm not sure Quarkus specific implementations are implemented in the right way because they are reflection based
2. Tests are missing

### CODE REVIEW

**Comment by radcortez** [comment]
@luca-bassoricci Thanks for the PR.

Maybe we can implement some of the changes in JBeret directly (change method visibilities), so we can use them here directly without requiring reflection. I had to do some of that work when we first wrote the extension. I'm sure @chengfang will help.

If this is a major blocker for you, I'm fine if we move with the reflection based approach while we do the work in JBeret, but we do need tests :)

Thanks!

**Comment by chengfang** [comment]
@luca-bassoricci  please file [JBeret JIRA](https://issues.redhat.com/issues/?jql=project%20%3D%20JBERET%20ORDER%20BY%20key%20DESC) issues for changes needed from [JBeret](https://github.com/jberet/jsr352), and we'll evaluate if it's suitable to include it.

**Comment by luca-bassoricci** [comment]
> @luca-bassoricci please file [JBeret JIRA](https://issues.redhat.com/issues/?jql=project%20%3D%20JBERET%20ORDER%20BY%20key%20DESC) issues for changes needed from [JBeret](https://github.com/jberet/jsr352), and we'll evaluate if it's suitable to include it.

JIRA issue: https://issues.redhat.com/browse/JBERET-561

**Comment by chengfang** [comment]
@luca-bassoricci  @radcortez  all 3 ScopedContextImpl classes have a static `INSTANCE` field (though not public). Can you use them instead of accessing their constructor? If so, we can add a getInstance() public method.

**Comment by chengfang** [comment]
After reading more about this PR, I'd like to suggest something slightly different:

Can `QuarkusJobScopedContextImpl` and the like extend jberet's `JobScopedContextImpl`?  This will avoid quite some method duplicatation, without the need to access the constructor.

Is the only purpose of accessing `getJobScopedBeans` is to use it in `destroy()` method? If so, the parent class in jberet can implement the `destroy()` method with the same logic accessing the jberet-private `getJobScopedBeans()`. So Quarkus side does not need to access `getJobScopedBeans()` at all.

Will the above work?

**Comment by luca-bassoricci** [comment]
> all 3 ScopedContextImpl classes have a static INSTANCE field (though not public). Can you use them instead of accessing their constructor? If so, we can add a getInstance() public method in jberet ScopedContextImpl classes.

> After reading more about this PR, I'd like to suggest something slightly different:
> 
> Can `QuarkusJobScopedContextImpl` and the like extend jberet's `JobScopedContextImpl`? This will avoid quite some method duplicatation, without the need to access the constructor.

As my personal opinion I prefer to access private `INSTANCE` fields via a static `getInstance()`; for example using a new class just for Quarkus (eg `QuarkusCDIExtension`) where scopes `INSTANCE`s are exposed using the a public  `getXXXScopeInstance()`; this is because I prefer to use delegation over extension.

> 
> Is the only purpose of accessing `getJobScopedBeans` is to use it in `destroy()` method? If so, the parent class in jberet can implement the `destroy()` method with the same logic accessing the jberet-private `getJobScopedBeans()`. So Quarkus side does not need to access `getJobScopedBeans()` at all.
> 
> Will the above work?

Yep, for now the solely usage is that.
I have only some concerns about my implementation of [`AlterableContext#destroy([Contextual](https://docs.oracle.com/javaee/7/api/javax/enterprise/context/spi/Contextual.html)<?> contextual)`](https://docs.oracle.com/javaee/7/api/javax/enterprise/context/spi/AlterableContext.html#destroy-javax.enterprise.context.spi.Contextual-) but I opened a review on my PR to let @radcortez have a look and give his opinion about this problem.

**Comment by chengfang** [comment]
Okay, based on the above discussion, here is my plan for jberet-side changes:
* add `public static getInstance()` method to jberet scoped context impl classes;
* add `public void destroy()` method to jberet scoped context impl classes, so quarkus-jberet classes can either inherit or delegate to it.
* constructor and `getScopedBeans()` methods in jberet will remain private (unchanged).

**Comment by chengfang** [comment]
https://github.com/jberet/jsr352/pull/232 is the jberet PR. Please review and let me know if anything else is needed.

Note that a single `destroy(Contextual<?>)` method is added. This method destroys a specific bean, but if called with `null`, it will behave the same as what you would want in `destroy()` -- to destroy all beans.

Also note that the latest quarkus `InjectableContext` interface has changed: there is now a `destroy(ContextState)` method replacing `destroy(Contextual<?>)`.

**Comment by luca-bassoricci** [comment]
> [jberet/jsr352#232](https://github.com/jberet/jsr352/pull/232) is the jberet PR. Please review and let me know if anything else is needed.
> 
> Note that a single `destroy(Contextual<?>)` method is added. This method destroys a specific bean, but if called with `null`, it will behave the same as what you would want in `destroy()` -- to destroy all beans.
> 
> Also note that the latest quarkus `InjectableContext` interface has changed: there is now a `destroy(ContextState)` method replacing `destroy(Contextual<?>)`.

I've seen your PR and LGTM, but I have also seen `InjectableContext` changes so I need more time to investigate.
I'll update you all ASAP.

**Comment by luca-bassoricci** [comment]
@chengfang 

> > [jberet/jsr352#232](https://github.com/jberet/jsr352/pull/232) is the jberet PR. Please review and let me know if anything else is needed.
> > Note that a single `destroy(Contextual<?>)` method is added. This method destroys a specific bean, but if called with `null`, it will behave the same as what you would want in `destroy()` -- to destroy all beans.
> > Also note that the latest quarkus `InjectableContext` interface has changed: there is now a `destroy(ContextState)` method replacing `destroy(Contextual<?>)`.
> 
> I've seen your PR and LGTM, but I have also seen `InjectableContext` changes so I need more time to investigate. I'll update you all ASAP.

@chengfang 
I suppose you can merge your PR.

**Comment by chrisschauer** [comment]
Any progress here? I stumbled over the same issue with JobScoped beans. 
Thanks

**Comment by radcortez** [comment]
Unsure. This PR has been stale for some time. @luca-bassoricci do you intend to keep working on this? Thanks!

**Comment by luca-bassoricci** [comment]
I'll be happy to work on it but on my (short) spare time because I am a bit busy at work, sorry.
I pinged cheofang times ago but without response.
Codebase on my branch is still with Quarkus 2.x, so I need to upgrade to latest Quarkus and JBeret.
May we drop support for Quarkus 2.x, WDYT @radcortez ?

**Comment by radcortez** [comment]
> May we drop support for Quarkus 2.x, WDYT @radcortez ?

Sure.

**Comment by liweinan** [comment]
I'll check this.

**Comment by luca-bassoricci** [comment]
I have aligned with latest extension codebase and JBeret 2.2.0-SNAPSHOT (current main) where chengfang PR will be merged later
@radcortez build fails because jberet 2.2.0-SNAPSHOT dependency: IDK which is the right way to proceed in this case when using an external snapshot repository as dep.

**Comment by radcortez** [comment]
One technique that I use is to add a step into the ci builder to clone and build locally the snapshot dependency until we have a stable release.

**Comment by luca-bassoricci** [comment]
> > One technique that I use is to add a step into the ci builder to clone and build locally the snapshot dependency until we have a stable release.
> 
> That's fine. Done.

@radcortez I'm currently use the main branch of Jberet to build locally, but what about using chengfang:scope-impl-public-JBERET-561 instead? This should avoid @liweinan to merge the PR until we decide to remove the draft to this PR.
WDYT?

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -21,6 +21,15 @@ jobs:
     name: build with jdk ${{matrix.java}}
 
     steps:
+      - uses: actions/checkout@v4
+        name: checkout jberet snapshot
+        with:
+          repository: jberet/jsr352
```
Comment (liweinan) [inline .github/workflows/build.yml:27]: @luca-bassoricci Maybe you can directly use cheng's fork repo here.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -21,6 +21,15 @@ jobs:
...
+          path: jberet_jsr352-snapshot
+
+      - name: build and install jberet/jsr352 snapshot with maven
+        run: mvn compile install --file jberet_jsr352-snapshot/pom.xml
```
Comment (liweinan) [inline .github/workflows/build.yml:31]: @luca-bassoricci And then checkout this branch from his fork

**Comment by radcortez** [comment]
> @radcortez I'm currently use the main branch of Jberet to build locally, but what about using chengfang:scope-impl-public-JBERET-561 instead? This should avoid @liweinan to merge the PR until we decide to remove the draft to this PR. WDYT?

Yes, it's fine.

**Comment by luca-bassoricci** [comment]
@liweinan created a new branch based on jberet 2.x and I'm already using it; cheng's branch IMO can be dismissed because we don't support pre-jakarta code.
@liweinan no rush to merge your branch, we can use your branch until this PR is draft

**Comment by liweinan** [comment]
@luca-bassoricci Cool! :-) I'll finish doing TCK tests today but I guess do the release need sometime, anyway I'll finish working on this before next weekend.

In addition, just to confirm that you don't need the PR for JBeret `1.4.x` right? If that's true I'll just do the release for `2.1.x`.

**Comment by luca-bassoricci** [comment]
> In addition, just to confirm that you don't need the PR for JBeret `1.4.x` right? If that's true I'll just do the release for `2.1.x`.

Confirmed

 
> To be concise: If Quarkus side has no specific requirement, and if you can just use the [jberet/jsr352#432 (comment)](https://github.com/jberet/jsr352/pull/432#issuecomment-1857373023), I'll do a release `2.2.0.Final` from main branch, because branch `2.1` is for EAP8.

IDK if Quarkus has specific requirements, maybe @radcortez can answer.
As stated above I'm currently use your PR branch while in draft but if backport to 2.1 (https://github.com/jberet/jsr352/pull/432#issuecomment-1856172415)  we can decide to move to 2.1.x-stable as base anytime

**Comment by liweinan** [comment]
@luca-bassoricci Thanks for the confirmation! Okay I'll do the `2.2.0.Final` once TCK tests are done, and if it needs to be back-ported into branch `2.1` in the future please let me know.

**Comment by liweinan** [comment]
@luca-bassoricci I have released `2.2.0.Final` for use: https://github.com/jberet/jsr352/releases/tag/2.2.0.Final If there is any problem please let me know.

**Comment by luca-bassoricci** [comment]
> 
> If this is a major blocker for you, I'm fine if we move with the reflection based approach while we do the work in JBeret, but we do need tests :)
> 

About tests:
My idea is to import JBeret tests from https://github.com/jberet/jsr352/tree/main/test-apps/cdiScopes and add them to integration-test module instead of create a new set of tests from scratch.
jberet-scope-tests


@radcortez WDYT?

**Comment by radcortez** [comment]
> About tests: My idea is to import JBeret tests from https://github.com/jberet/jsr352/tree/main/test-apps/cdiScopes and add them to integration-test module instead of create a new set of tests from scratch. 
>
> @radcortez WDYT?

Sounds good.

**Comment by luca-bassoricci** [comment]
Tests added.
I have used JBeret test artifacts, create alternative beans because JBeret's ones are not `@Dependent` and copied *IT tests.
commons project contains a base abstract test class useful to integrate with JBeret test classes.
Let me know

**Comment by luca-bassoricci** [comment]
> Tests added. I have used JBeret test artifacts, create alternative beans because JBeret's ones are not `@Dependent` and copied *IT tests. `commons` project contains a base abstract test class useful to integrate with JBeret test classes. I had to add a reflection call [here](https://github.com/quarkiverse/quarkus-jberet/pull/123/files#diff-f004ce9e7a3d666465c079bb2ffd37e8f45c55477cee65934d745b50dde22091) but I don't think is a concern because it only a test and not runtime code. Let me know.

@radcortez any comment?

**Comment by radcortez** [comment]
> @radcortez any comment?

Yes, it is fine. Thank you.

So, is this complete? Or are we still missing something?

**Comment by luca-bassoricci** [comment]
From my POV all seems to be fine; for me a big concern was about the tests but after JBeret's tests passed I'm more confident.
I suppose draft status can be removed and PR merged.

**Comment by luca-bassoricci** [comment]
> > @radcortez any comment?
> 
> Yes, it is fine. Thank you.
> 
> So, is this complete? Or are we still missing something?

From my POV all seems to be fine; for me a big concern was about the tests but after JBeret's tests passed I'm more confident.
I suppose draft status can be removed and PR merged.

**Comment by liweinan** [comment]
btw JBeret `2.2.0.Final` has been merged into WildFly main branch: https://github.com/wildfly/wildfly/pull/17519#event-11463054652

File: core/deployment/src/main/java/io/quarkiverse/jberet/deployment/JBeretProcessor.java [hunk 3]
```diff
@@ -99,28 +110,37 @@ public class JBeretProcessor {
     private static final DotName JOB = DotName.createSimple(Job.class);
 
     @BuildStep
-    public void registerExtension(BuildProducer<FeatureBuildItem> feature, BuildProducer<CapabilityBuildItem> capability) {
+    public void registerExtension(BuildProducer<FeatureBuildItem> feature,
```
Comment (radcortez) [inline core/deployment/src/main/java/io/quarkiverse/jberet/deployment/JBeretProcessor.java:113]: It seems that in this class, there are a bunch of changes that are only code style changes and not related to the actual feature. Any chance we can revert these and keep only the relevant changes?

**Review by radcortez — CHANGES_REQUESTED** [review]

**Review by radcortez — APPROVED** [review]
Thank you for all the work. Really appreciate it :)

**Pull request merged**